build.rs

```diff
@@ -1,16 +1,70 @@
 extern crate pkg_config;
 
+#[cfg(feature = "use-bindgen")]
+extern crate bindgen;
+
+#[allow(unused_variables)]
 fn main() {
-    if let Err(e) = pkg_config::probe_library("alsa") {
-        match e {
-            pkg_config::Error::Failure { .. } => panic! (
-                "Pkg-config failed - usually this is because alsa development headers are not installed.\n\n\
-                For Fedora users:\n# dnf install alsa-lib-devel\n\n\
-                For Debian/Ubuntu users:\n# apt-get install libasound2-dev\n\n\
-                pkg_config details:\n{}",
-                e
-            ),
-            _ => panic!("{}", e)
-        }
-    }
+    let alsa_library = match pkg_config::probe_library("alsa") {
+        Err(e) => {
+            match e {
+                pkg_config::Error::Failure { .. } => panic! (
+                    "Pkg-config failed - usually this is because alsa development headers are not installed.\n\n\
+                    For Fedora users:\n# dnf install alsa-lib-devel\n\n\
+                    For Debian/Ubuntu users:\n# apt-get install libasound2-dev\n\n\
+                    pkg_config details:\n{}",
+                    e
+                ),
+                _ => panic!("{}", e)
+            }
+        },
+        Ok(l) => l
+    };
+
+    #[cfg(feature = "use-bindgen")]
+    generate_bindings(&alsa_library);
+}
+
+#[cfg(feature = "use-bindgen")]
+fn generate_bindings(alsa_library: &pkg_config::Library) {
+    use std::env;
+    use std::path::PathBuf;
+
+    println!("cargo:rerun-if-env-changed=FORCE_REGENERATE_BINDINGS");
+
+    let clang_include_args = alsa_library.include_paths.iter().map(|include_path| {
+        format!(
+            "-I{}",
+            include_path.to_str().expect("include path was not UTF-8")
+        )
+    });
```

Comment:
This bit is kinda inefficient... the pkg-config crate parses out the `-I/usr/include` etc. paths for us from the output of the pkg-config CLI util, and then we have to map them back into the `-I` format to pass along to clang.

However, we get a bunch of extra helpful stuff by using the pkg-config crate, so I think it's probably the best thing to do.

Reply:
Agreed